nameko/testing/services.py

```diff
@@ -48,93 +48,157 @@ def entrypoint_hook(container, method_name, context_data=None):
 
-    Usage::
+        worker_ctx (WorkerContext): WorkerContext of the entrypoint call.
 
-        container = ServiceContainer(ExampleService, config)
-        with entrypoint_waiter(container, 'example_handler'):
-            ...  # e.g. rpc call that will result in handler being called
-    """
+        result (object): The result, if any, that the entrypoint returned.
+
+        exc_info (tuple): Tuple as returned by `sys.exc_info` if the
+            entrypoint raised an exception, otherwise `None`.
 
-    waiter = EntrypointWaiter(entrypoint)
-    if not get_extension(container, Entrypoint, method_name=entrypoint):
+
+    The `timeout` argument specifies the maximum number of seconds the
+    `entrypoint_waiter` should wait before exiting. It can be disabled by
+    passing `None`.
+
+    **Usage**
+
+    ::
+        class Service(object):
+            name = "service"
+
+            @event_handler('srcservice', 'eventtype')
+            def handle_event(self, msg):
+                return msg
+
+        container = ServiceContainer(Service, config)
+        container.start()
+
+        # basic
+        with entrypoint_waiter(container, 'handle_event'):
+            ...  # action that dispatches event
+
+        # giving access to the result
+        with entrypoint_waiter(container, 'handle_event') as result:
+            ...  # action that dispatches event
+        res = result.get()
+
+        # with custom timeout
+        with entrypoint_waiter(container, 'handle_event', timeout=5):
+            ...  # action that dispatches event
+
+        # with callback that waits until entrypoint stops raising
+        def callback(worker_ctx, result, exc_info):
+            if exc_info is None:
+                return True
+
+        with entrypoint_waiter(container, 'handle_event', callback=callback):
+            ...  # action that dispatches event
+
+    """
+    if not get_extension(container, Entrypoint, method_name=method_name):
         raise RuntimeError("{} has no entrypoint `{}`".format(
-            container.service_name, entrypoint))
-    if get_extension(container, EntrypointWaiter, entrypoint=entrypoint):
-        raise RuntimeError("Waiter already registered for {}".format(
-            entrypoint))
-
-    # can't mess with dependencies while container is running
-    wait_for_worker_idle(container)
-    container.dependencies.add(waiter)
-
-    try:
-        yield
-        exc = waiter.Timeout(
-            "Entrypoint {}.{} failed to complete within {} seconds".format(
-                container.service_name, entrypoint, timeout)
-        )
-        with eventlet.Timeout(timeout, exception=exc):
-            waiter.wait()
-    finally:
-        wait_for_worker_idle(container)
-        container.dependencies.remove(waiter)
+            container.service_name, method_name))
+
+    waiter_callback = callback
+    waiter_result = WaitResult()
+
+    def on_worker_result(worker_ctx, result, exc_info):
+        complete = False
```

Comment:
I see your point.

We _could_ simplify the entrypoint waiter callback to be the same as `wait_for_call`, so they'd both be:

``` python
def cb(args, kwargs, result, exc_info):
   pass
```

But what you lose is the ability to interact with the `worker_ctx` in the entrypoint waiter callback, which contains lots of useful information about the call beyond the arguments.

The "worker context" concept is related to the execution of an entrypoint, whereas the `wait_for_call` helper really is just about waiting for a normal method execution. It's unfortunate that they have such similar signatures.

An alternative way to differentiate the APIs would be to attach the result or exception to the `worker_ctx` object rather than passing them around separately. Then the signature of the `entrypoint_waiter` callback would be simply:

``` python
def cb(worker_ctx):
    pass
```